### PR TITLE
Adds token refresh logic

### PR DESCRIPTION
**Pull request by loreto**
## Summary
Adds token refresh logic

## How was it tested?
I'm actually unsure how best to test this other than going through the auth flow, waiting for my token to expire, and running again. Any ideas on how best to shortcut that process? Is there a way to purposefully give myself an expired token on the first try?

### CODE REVIEW

**Comment by mikeland73** [comment]
> I'm actually unsure how best to test this other than going through the auth flow, waiting for my token to expire, and running again. Any ideas on how best to shortcut that process? Is there a way to purposefully give myself an expired token on the first try?

* You can change the expiration to be super short. 
* You can also just call refresh and inspect that the tokens changed and you got new expirations.
* Add `ValidAlwaysExpiredForTesting` function and use it to test

File: pkg/sandbox/auth/auth.go [hunk 1]
```diff
@@ -61,14 +64,58 @@ func (c *Client) LogoutFlow() error {
...
 		return nil, false
 	}
+
+	// Refresh if the token is no longer valid:
+	if !tok.Valid() {
+		tok = c.refresh(tok)
+		if !tok.Valid() {
+			return nil, false
+		}
+		return tok, true
```
Comment (mikeland73) [inline pkg/sandbox/auth/auth.go:77]: no needed

File: pkg/sandbox/auth/auth.go [hunk 2]
```diff
@@ -61,14 +64,58 @@ func (c *Client) LogoutFlow() error {
...
+	}
+
 	return tok, true
 }
 
-func (c *Client) RefreshSession() *session.Token {
-	panic("refresh session not implemented")
+func (c *Client) refresh(tok *session.Token) *session.Token {
+	ctx := context.Background()
```
Comment (mikeland73) [inline pkg/sandbox/auth/auth.go:84]: nit, pass it in.

File: pkg/sandbox/auth/auth.go [hunk 3]
```diff
@@ -61,14 +64,58 @@ func (c *Client) LogoutFlow() error {
...
+	}
+
+	// Refresh logic:
+	tokenSource := conf.TokenSource(ctx, &tok.Token)
+	newToken, err := tokenSource.Token()
+	if err != nil {
+		return tok
+	}
+
+	if newToken.AccessToken != tok.AccessToken {
```
Comment (mikeland73) [inline pkg/sandbox/auth/auth.go:110]: Is this check needed?

**Review by mikeland73 — APPROVED** [review]

**Review by mikeland73 — APPROVED** [review]

File: pkg/sandbox/auth/auth.go [hunk 4]
```diff
@@ -59,16 +62,60 @@ func (c *Client) LogoutFlow() error {
...
-	panic("refresh session not implemented")
+func (c *Client) refresh(
+	ctx context.Context,
+	tok *session.Token,
+) *session.Token {
+	if tok == nil {
+		return nil
+	}
+
+	// TODO: figure out how to share oidc provider and outh2 client
```
Comment (LucilleH) [inline pkg/sandbox/auth/auth.go:90]: typo: `outh2`

**Review by LucilleH — APPROVED** [review]

**Comment by mikeland73** [comment]
I tested by:

* Had an expired id token
* Ran `envsec auth whoami` which triggered refresh
* Printed correct info and extended life of token

**Pull request merged**